transport/src/main/java/io/netty/channel/pool/DefaultChannelPoolKey.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.pool;
+
+import io.netty.channel.EventLoop;
+
+import java.net.SocketAddress;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Default implementation of a {@link ChannelPoolKey}.
+ */
+public final class DefaultChannelPoolKey implements ChannelPoolKey {
+
+    private final SocketAddress remoteAddress;
+    private final EventLoop loop;
+    private final int hash;
+
+    /**
+     * @see {@link #DefaultChannelPoolKey(SocketAddress, EventLoop)}
+     */
+    public DefaultChannelPoolKey(SocketAddress remoteAddress) {
+        this(remoteAddress, null);
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param remoteAddress     The {@link SocketAddress} of the remote peer.
+     * @param loop              The {@link EventLoop} that should be used or {@code null} if not specified.
+     */
+    public DefaultChannelPoolKey(SocketAddress remoteAddress, EventLoop loop) {
+        this.remoteAddress = checkNotNull(remoteAddress, "address");
+        this.loop = loop;
+        hash = (loop == null) ? remoteAddress.hashCode() : (37 * remoteAddress.hashCode() + loop.hashCode());
+    }
+
+    @Override
+    public SocketAddress remoteAddress() {
+        return remoteAddress;
+    }
+
+    @Override
+    public EventLoop eventLoop() {
+        return loop;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof ChannelPoolKey)) {
+            return false;
+        }
+        if (obj == this) {
+            return true;
+        }
+        ChannelPoolKey key = (ChannelPoolKey) obj;
+        if (!remoteAddress.equals(key.remoteAddress())) {
+            return false;
+        }
+        if (loop != null && !loop.equals(key.eventLoop())) {
```

Comment:
@normanmaurer - FYI. I may be mis-interpreting the output of `-XX:+PrintCompilation` but I don't think JIT optimizes this. `equalsMultipleIfs` is the code you have here and `equalsCompactIf` is the code with the last three if statements combined with `||`.

```
    136   62       3       com.test::equalsMultipleIfs (92 bytes)
    137   63       3       com.test::equalsCompactIf (88 bytes)
    138   67       4       com.test::equalsMultipleIfs (92 bytes)
    138   68       4       com.test::equalsCompactIf (88 bytes)
    140   62       3       com.test::equalsMultipleIfs (92 bytes)   made not entrant
    141   63       3       com.test::equalsCompactIf (88 bytes)   made not entrant
```